File: src/agent_memory/controller/trust.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class TrustScorer:
    recency_weight: float = 0.15
    corroboration_weight: float = 0.15
    contradiction_weight: float = 0.2
    source_weight: float = 0.5

    def score(
        self,
        *,
        source_reliability: float,
        corroboration_count: int = 0,
        contradiction_count: int = 0,
        age_days: float = 0.0,
    ) -> float:
        recency_bonus = max(0.0, 1.0 - min(age_days, 90.0) / 90.0)
        corroboration_bonus = min(corroboration_count, 5) / 5.0
        contradiction_penalty = min(contradiction_count, 5) / 5.0
        raw_score = (
            source_reliability * self.source_weight
            + recency_bonus * self.recency_weight
            + corroboration_bonus * self.corroboration_weight
            - contradiction_penalty * self.contradiction_weight
        )
        return max(0.0, min(1.0, raw_score))

File: go-server/internal/controller/trust.go
package controller

type TrustScorer struct {
	RecencyWeight       float64
	CorroborationWeight float64
	ContradictionWeight float64
	SourceWeight        float64
}

func NewTrustScorer() TrustScorer {
	return TrustScorer{
		RecencyWeight:       0.15,
		CorroborationWeight: 0.15,
		ContradictionWeight: 0.2,
		SourceWeight:        0.5,
	}
}

func (scorer TrustScorer) Score(sourceReliability float64, corroborationCount int, contradictionCount int, ageDays float64) float64 {
	recencyBonus := 1.0 - min(ageDays, 90.0)/90.0
	if recencyBonus < 0 {
		recencyBonus = 0
	}
	corroborationBonus := min(float64(corroborationCount), 5) / 5.0
	contradictionPenalty := min(float64(contradictionCount), 5) / 5.0
	rawScore := sourceReliability*scorer.SourceWeight +
		recencyBonus*scorer.RecencyWeight +
		corroborationBonus*scorer.CorroborationWeight -
		contradictionPenalty*scorer.ContradictionWeight
	if rawScore < 0 {
		return 0
	}
	if rawScore > 1 {
		return 1
	}
	return rawScore
}

func min(left float64, right float64) float64 {
	if left < right {
		return left
	}
	return right
}
